packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,279 @@
+  const sortedActionsByMostRecent = (a: rawUserAction, b: rawUserAction) =>
+    Number(b.time) - Number(a.time);
+
+  return actions.sort(sortedActionsByMostRecent);
+};
+
+export const filterActions = (
+  actions: UserAction[],
+  excludedCustomActions: string[]
+): UserAction[] => {
+  const filteredActions = actions.filter((action) => {
+    let shouldExcludeCustomAction = false;
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    if (eventType && eventValue) {
+      shouldExcludeCustomAction =
+        excludedCustomActions.includes(eventType) ||
+        excludedCustomActions.includes(eventValue);
+    }
+    return (
+      action.actionType !== UserActionType.CUSTOM || shouldExcludeCustomAction
+    );
+  });
+  return filteredActions;
+};
+
+export const buildUserActionFromRawAction = (
+  rawActions: rawUserAction[]
+): UserAction[] => {
+  const mappedUserActions = rawActions.map((rawAction) => {
+    const actionData = JSON.parse(rawAction.value);
+    return {
+      actionType: rawAction.name as UserActionType,
+      timestamp: rawAction.time,
+      eventData: {
+        type: actionData.event_type,
+        value: actionData.event_value,
+      },
+      cause: actionData.cause,
+      searchHub: actionData.origin_level_1,
+      document: {
+        title: actionData.title,
+        uriHash: actionData.uri_hash,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdKey: actionData.c_contentidkey,
+        // eslint-disable-next-line @cspell/spellchecker
+        contentIdValue: actionData.c_contentidvalue,
+      },
+      query: actionData.query_expression,
+    };
+  });
+  return mappedUserActions;
+};
+
+export const splitActionsIntoTimelineSessions = (
+  actions: UserAction[],
+  ticketCreationDate: string
+): UserActionTimeline => {
+  const returnTimeline: UserActionTimeline = {
+    precedingSessions: [],
+    session: undefined,
+    followingSessions: [],
+    caseSessionFound: false,
+  };
+
+  let currentSession: UserSession = {
+    start: actions[0].timestamp,
+    end: actions[0].timestamp,
+    actions: [],
+  };
+
+  actions.forEach((action) => {
+    if (isPartOfTheSameSession(action, currentSession.end)) {
+      currentSession.actions.push(action);
+      currentSession.start = action.timestamp;
+      return;
+    }
+
+    handleSessionEnd(currentSession, returnTimeline, ticketCreationDate);
+
+    currentSession = {
+      start: action.timestamp,
+      end: action.timestamp,
+      actions: [action],
+    };
+  });
+
+  // Inserting ticket creation action in current session
+  const caseCreationSessionActions: UserAction[] | [] =
+    insertTicketCreationActionInSession(
+      returnTimeline.session,
+      ticketCreationDate
+    );
+
+  const caseCreationSession = createCaseCreationSession(
+    caseCreationSessionActions
+  );
+  returnTimeline.session = caseCreationSession;
```

Comment:
Please see my previous comment it would allow use to add this, 
the problem here is that you are creating functions that returns the new value instead of directly doing the modification directly on the provided object, that's why it seems comlicated and not very readable.